src/analysis/specific_limits.lean

```diff
@@ -195,6 +195,61 @@ lemma geom_le {u : ℕ → ℝ} {c : ℝ} (hc : 0 ≤ c) (n : ℕ) (h : ∀ k <
   c ^ n * u 0 ≤ u n :=
 by refine (monotone_mul_left_of_nonneg hc).seq_le_seq n _ _ h; simp [pow_succ, mul_assoc, le_refl]
 
+/-- For any natural `k` and a real `r > 1` we have `n ^ k = o(r ^ n)` as `n → ∞`. -/
+lemma is_o_pow_const_const_pow_of_one_lt {R : Type*} [normed_ring R] (k : ℕ) {r : ℝ} (hr : 1 < r) :
+  is_o (λ n, n ^ k : ℕ → R) (λ n, r ^ n) at_top :=
+begin
+  have : tendsto (λ x : ℝ, x ^ k) (𝓝[Ioi 1] 1) (𝓝 1),
+    from ((continuous_id.pow k).tendsto' (1 : ℝ) 1 (one_pow _)).mono_left inf_le_left,
+  obtain ⟨r' : ℝ, hr' : r' ^ k < r, h1 : 1 < r'⟩ :=
+    ((this.eventually (gt_mem_nhds hr)).and self_mem_nhds_within).exists,
+  have h0 : 0 ≤ r' := zero_le_one.trans h1.le,
+  suffices : is_O _ (λ n : ℕ, (r' ^ k) ^ n) at_top,
+    from this.trans_is_o (is_o_pow_pow_of_lt_left (pow_nonneg h0 _) hr'),
+  conv in ((r' ^ _) ^ _) { rw [← pow_mul, mul_comm, pow_mul] },
+  suffices : ∀ n : ℕ, ∥(n : R)∥ ≤ (r' - 1)⁻¹ * ∥(1 : R)∥ * ∥r' ^ n∥, from (is_O_of_le' _ this).pow _,
+  intro n, rw mul_right_comm,
+  refine n.norm_cast_le.trans (mul_le_mul_of_nonneg_right _ (norm_nonneg _)),
+  simpa [div_eq_inv_mul, real.norm_eq_abs, abs_of_nonneg h0] using n.cast_le_pow_div_sub h1
+end
+
+/-- For a real `r > 1` we have `n = o(r ^ n)` as `n → ∞`. -/
+lemma is_o_coe_const_pow_of_one_lt {R : Type*} [normed_ring R] {r : ℝ} (hr : 1 < r) :
+  is_o (coe : ℕ → R) (λ n, r ^ n) at_top :=
+by simpa only [pow_one] using is_o_pow_const_const_pow_of_one_lt 1 hr
+
+/-- If `∥r₁∥ < r₂`, then for any naturak `k` we have `n ^ k r₁ ^ n = o (r₂ ^ n)` as `n → ∞`. -/
+lemma is_o_pow_const_mul_const_pow_const_pow_of_norm_lt {R : Type*} [normed_ring R] (k : ℕ)
+  {r₁ : R} {r₂ : ℝ} (h : ∥r₁∥ < r₂) :
+  is_o (λ n, n ^ k * r₁ ^ n : ℕ → R) (λ n, r₂ ^ n) at_top :=
+begin
+  by_cases h0 : r₁ = 0,
+  { refine (is_o_zero _ _).congr' (mem_at_top_sets.2 $ ⟨1, λ n hn, _⟩) eventually_eq.rfl,
+    simp [zero_pow (zero_lt_one.trans_le hn), h0] },
+  rw [← ne.def, ← norm_pos_iff] at h0,
+  have A : is_o (λ n, n ^ k : ℕ → R) (λ n, (r₂ / ∥r₁∥) ^ n) at_top,
+    from is_o_pow_const_const_pow_of_one_lt k ((one_lt_div h0).2 h),
+  suffices : is_O (λ n, r₁ ^ n) (λ n, ∥r₁∥ ^ n) at_top,
+    by simpa [div_mul_cancel _ (pow_pos h0 _).ne'] using A.mul_is_O this,
+  exact is_O.of_bound 1 (by simpa using eventually_norm_pow_le r₁)
+end
+
+lemma tendsto_pow_const_div_const_pow_of_one_lt (k : ℕ) {r : ℝ} (hr : 1 < r) :
+  tendsto (λ n, n ^ k / r ^ n : ℕ → ℝ) at_top (𝓝 0) :=
+(is_o_pow_const_const_pow_of_one_lt k hr).tendsto_0
+
+/-- If `|r| < 1`, then `n ^ k r ^ n` tends to zero for any natural `k`. -/
+lemma tendsto_pow_const_mul_const_pow_of_abs_lt_one (k : ℕ) {r : ℝ} (hr : abs r < 1) :
+  tendsto (λ n, n ^ k * r ^ n : ℕ → ℝ) at_top (𝓝 0) :=
+begin
+  by_cases h0 : r = 0,
+    from tendsto_const_nhds.congr'
+      (mem_at_top_sets.2 ⟨1, λ n hn, by simp [zero_lt_one.trans_le hn, h0]⟩),
```

Comment:
the indentation is strange here, since you're just closing one of the two subgoals created by `by_cases`.